contracts/staking/fuse/FuseStaking.sol

```diff
@@ -0,0 +1,743 @@
+				address(this),
+				validators[i]
+			);
+			total += cur;
+		}
+		return total;
+	}
+
+	function _checkIfCalledOnceInDayAndReturnDay() internal returns(uint256) {
+		uint256 curDay = ubiScheme.currentDay();
+		require(
+			curDay != lastDayCollected,
+			"can collect only once in a ubi cycle"
+		);
+		lastDayCollected = curDay;
+		return curDay;
+	}
+
+	function _getLastRewardPerToken() internal view returns(uint256) {
+		return globalStakingVariablesHistory.length > 0 ? globalStakingVariablesHistory[globalStakingVariablesHistory.length - 1].rewardPerToken : 0;
+	}
+
+	function _getLastRewardRate() internal view returns(uint256) {
+		return globalStakingVariablesHistory.length > 0 ? globalStakingVariablesHistory[globalStakingVariablesHistory.length - 1].rewardRate : 0;
+	}
+
+	function _getLastTotalFinalizedSupply() internal view returns(uint256) {
+		return globalStakingVariablesHistory.length > 0 ? globalStakingVariablesHistory[globalStakingVariablesHistory.length - 1].totalFinalizedSupply : 0;
+	}
+
+	function _getRewardPerTokenPerUser(address _account) internal view returns(uint256) {
+		if (collectUBIInterestCallTimes.length > 1) {
+			uint256 rewardPerTokenPerUser = 0;
+			for (uint256 i = 1; i < collectUBIInterestCallTimes.length; i++) {
+				if (collectUBIInterestCallTimeIdxToUserStakeTime[i][_account] == 0) continue;
+				rewardPerTokenPerUser +=
+					_rewardPerToken(
+						globalStakingVariablesHistory[i].rewardPerToken,
+						collectUBIInterestCallTimes[i] - collectUBIInterestCallTimeIdxToUserStakeTime[i - 1][_account],
+						globalStakingVariablesHistory[i].rewardRate,
+						globalStakingVariablesHistory[i].totalFinalizedSupply
+					);
+			}
+			return rewardPerTokenPerUser;
+		} else {
+			return 0;
+		}
+	}
+
+	function earned(address account) public view returns (uint256) {
+		return pendingStakes[account] * (_getRewardPerTokenPerUser(account) - userRewardPerTokenPaid[account]) / PRECISION;
+	}
+
+	function _getReward(address _to) internal {
+		uint256 reward = earned(_to);
+		if (reward > 0) {
+				userRewardPerTokenPaid[_to] = _getRewardPerTokenPerUser(_to);
+				payable(_to).transfer(reward);
+		}
+	}
+
+	function getReward() external nonReentrant {
+			_getReward(msg.sender);
+	}
+
+	function _distributeGivebackAndQueryOracle(uint256 _amount) internal {
+		if (_amount == 0) return;
+		uint256[] memory faucetAddresses = spendingRateOracle.getFaucets();
+		for (uint256 i = 0; i < faucetAddresses.length; i++) {
+			address faucetToken = spendingRateOracle.getFaucetTokenAddress(faucetAddresses[i]);
+			uint256 targetBalance = spendingRateOracle.getFaucetTargetBalance(faucetAddresses[i]);
+			if (faucetToken == address(0)) {
+				uint256 balancesDifference;
+				if (faucetToken.balance < targetBalance) {
+					balancesDifference = targetBalance - faucetToken.balance;
+					if (_amount < balancesDifference) break;
+					_amount -= balancesDifference;
+					payable(faucetAddresses[i]).transfer(balancesDifference);
```

Comment:
how do you know we have enough?

Reply:
If we havent, It'll just do nothing and I planned to added some event that will say to us that break was occured (in if with break), I'll account that after the prototype.

Reply:
1. nothing are you sure? wouldnt transfer revert?
2. but we dont need nothing to happen, we need the max available to be transfered

Reply:
1. it'll just return false, but I think you right, I'll surround it by require func
2. Ok will do that

Reply:
1. do we want it to revert? or do we wantt it to succeed? analyzee and explain your answer